File: coding/src/com/algos/li16_design/revised/L460_LFUCache.java
package  com.algos.li16_design.revised;

import java.util.HashMap;
import java.util.Map;

public class L460_LFUCache {
    final int capacity;
    int curSize;
    int minFreq;
    Map<Integer, DLLNode> cache;
    Map<Integer, DoubleLinkedList> freqMap;

    public L460_LFUCache(int capacity) {
        this.capacity = capacity;
        this.curSize = 0;
        this.minFreq = 0;

        this.cache = new HashMap<>();
        this.freqMap = new HashMap<>();
    }

    public int get(int key) {
        DLLNode curNode = cache.get(key);
        if (curNode == null) {
            return -1;
        }
        updateNode(curNode);
        return curNode.val;
    }

    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            DLLNode curNode = cache.get(key);
            curNode.val = value;
            updateNode(curNode);
        }
        else {
            curSize++;
            if (curSize > capacity) {
                // get minimum frequency list
                DoubleLinkedList minFreqList = freqMap.get(minFreq);
                cache.remove(minFreqList.tail.prev.key);
                minFreqList.removeNode(minFreqList.tail.prev);
                curSize--;
            }
            // reset min frequency to 1 because of adding new node
            minFreq = 1;
            DLLNode newNode = new DLLNode(key, value);

            // get the list with frequency 1, and then add new node into the list, as well as into LFU cache
            DoubleLinkedList curList = freqMap.getOrDefault(1, new DoubleLinkedList());
            curList.addNode(newNode);
            freqMap.put(1, curList);
            cache.put(key, newNode);
        }
    }

    public void updateNode(DLLNode curNode) {
        int curFreq = curNode.frequency;
        DoubleLinkedList curList = freqMap.get(curFreq);
        curList.removeNode(curNode);

        // if current list the the last list which has lowest frequency and current node is the only node in that list
        // we need to remove the entire list and then increase min frequency value by 1
        if (curFreq == minFreq && curList.listSize == 0) {
            minFreq++;
        }

        curNode.frequency++;
        // add current node to another list has current frequency + 1,
        // if we do not have the list with this frequency, initialize it
        DoubleLinkedList newList = freqMap.getOrDefault(curNode.frequency, new DoubleLinkedList());
        newList.addNode(curNode);
        freqMap.put(curNode.frequency, newList);
    }
}

class DLLNode {
    int key;
    int val;
    int frequency;
    DLLNode prev;
    DLLNode next;

    public DLLNode(int key, int val) {
        this.key = key;
        this.val = val;
        this.frequency = 1;
    }
}

class DoubleLinkedList {
    int listSize;
    DLLNode head;
    DLLNode tail;
    public DoubleLinkedList() {
        this.listSize = 0;
        this.head = new DLLNode(0, 0);
        this.tail = new DLLNode(0, 0);
        head.next = tail;
        tail.prev = head;
    }

    public void addNode(DLLNode curNode) {
        DLLNode nextNode = head.next;
        curNode.next = nextNode;
        curNode.prev = head;
        head.next = curNode;
        nextNode.prev = curNode;
        listSize++;
    }

    public void removeNode(DLLNode curNode) {
        DLLNode prevNode = curNode.prev;
        DLLNode nextNode = curNode.next;
        prevNode.next = nextNode;
        nextNode.prev = prevNode;
        listSize--;
    }
}

File: dsa/topics/16_design/python/l460_lfucache.py
# Problem: https://leetcode.com/problems/lfu-cache/
# Idea:
# LFU = Least Frequently Used
#
# Data Structures:
# 1) cache        : key -> DLLNode
# 2) freqMap      : frequency -> DoubleLinkedList (nodes with same frequency)
# 3) minFreq      : minimum frequency currently present
#
# get(key):
#   - If not present, return -1
#   - Update node frequency
#
# put(key, value):
#   - If key exists, update value and frequency
#   - Else insert new node
#   - If capacity exceeded, evict LRU node from minFreq list
#
# Time Complexity: O(1) for get and put
# Space Complexity: O(capacity)

class LFUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.curSize = 0
        self.minFreq = 0

        self.cache = {}      # key -> DLLNode
        self.freqMap = {}    # freq -> DoubleLinkedList

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1

        node = self.cache[key]
        self.updateNode(node)
        return node.val

    def put(self, key: int, value: int) -> None:
        if self.capacity == 0:
            return

        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.updateNode(node)
        else:
            self.curSize += 1

            if self.curSize > self.capacity:
                # Remove LRU node from the minimum frequency list
                minFreqList = self.freqMap[self.minFreq]
                lruNode = minFreqList.tail.prev
                minFreqList.removeNode(lruNode)
                del self.cache[lruNode.key]
                self.curSize -= 1

            # Insert new node with frequency = 1
            self.minFreq = 1
            newNode = DLLNode(key, value)

            freqList = self.freqMap.get(1, DoubleLinkedList())
            freqList.addNode(newNode)
            self.freqMap[1] = freqList
            self.cache[key] = newNode

    def updateNode(self, node):
        curFreq = node.frequency
        curList = self.freqMap[curFreq]
        curList.removeNode(node)

        # If this was the only node with minFreq, increase minFreq
        if curFreq == self.minFreq and curList.listSize == 0:
            self.minFreq += 1

        node.frequency += 1
        newList = self.freqMap.get(node.frequency, DoubleLinkedList())
        newList.addNode(node)
        self.freqMap[node.frequency] = newList


class DLLNode:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.frequency = 1
        self.prev = None
        self.next = None


class DoubleLinkedList:
    def __init__(self):
        self.listSize = 0
        self.head = DLLNode(0, 0)
        self.tail = DLLNode(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def addNode(self, node):
        nextNode = self.head.next
        node.next = nextNode
        node.prev = self.head
        self.head.next = node
        nextNode.prev = node
        self.listSize += 1

    def removeNode(self, node):
        prevNode = node.prev
        nextNode = node.next
        prevNode.next = nextNode
        nextNode.prev = prevNode
        self.listSize -= 1
